presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageReader.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.pagefile;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.page.PagesSerde;
+import com.facebook.presto.spi.page.SerializedPage;
+import io.airlift.slice.InputStreamSliceInput;
+import io.airlift.slice.SliceInput;
+import org.apache.hadoop.fs.FSDataInputStream;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static com.facebook.presto.spi.page.PagesSerdeUtil.readSerializedPage;
+import static com.facebook.presto.spi.page.PagesSerdeUtil.serializedPageWrittenBytes;
+import static java.util.Objects.requireNonNull;
+
+public class PageFilePageReader
+        implements Iterator<Page>
+{
+    private final PagesSerde pagesSerde;
+    private final SliceInput input;
+    private final long readLength;
+    private long readBytes;
+
+    PageFilePageReader(
+            long readStart,
+            long readLength,
+            FSDataInputStream inputStream,
```

Comment:
How about passing an already positioned and limited input stream? (doing `inputStream.seek(readStart);` and `ByteStreams#limit` outside)? If you do that then most likely you won't even need this class.

Reply:
Keep using `InputStreamSliceInput` to reuse function `readSerializedPage(SliceInput)`

Reply:
There is one copy to internal slice of InputStreamSliceInput,  the advantage of doing it this way is to lower the count of real IO operation? For integer and small page, read from buffer is more efficient.

Reply:
@aweisberg Currently block recoding relies on SliceInput. There's only one existing `InputStreamSliceInput` implementation, and unfortunately it is not very efficient. We had a discussion around that with Vic. I think we all agree that it has to be optimized. What do you think about doing it as a follow up?

Reply:
It doesn't reduce the number of IO operations and will increase it in some cases. Basically if the size of the remaining bytes to read is > than the internal buffer size (or size > some multiple) it doesn't make sense to copy to the internal buffer. Pass the argument array to the wrapped stream along with the correct offset and length and do it directly.

Agree it should be a followup. It's not even a change we would make to this repo.